src/cosim/serialization.cpp

```diff
@@ -75,25 +86,507 @@ namespace
+            auto ctx = reinterpret_cast<cbor_reader_ctx*>(_ctx);
+            add_value(ctx, value);
+            check_map_end(ctx);
+        }
+
+        static void cbor_read_byte_string(void * _ctx, cbor_data data, uint64_t len)
+        {
+            auto ctx = reinterpret_cast<cbor_reader_ctx*>(_ctx);
+            auto value = std::string(reinterpret_cast<const char*>(data), len);
+            add_value(ctx, value);
+            check_map_end(ctx);
+        }
+
+        template<typename T>
+        static void cbor_read_int(void* _ctx, T value)
+        {
+            auto ctx = reinterpret_cast<cbor_reader_ctx*>(_ctx);
+
+            // If tag is defined, read the value as signed integer.
+            if (ctx->tag) {
+                add_value(ctx, static_cast<typename ::serialization::int_type<T>::type>(value));
+                ctx->tag = std::nullopt;
+            } else {
+                add_value(ctx, value);
+            }
+
+            check_map_end(ctx);
+        }
+
+        static void cbor_read_int(void* _ctx, uint8_t value)
+        {
+            auto ctx = reinterpret_cast<cbor_reader_ctx*>(_ctx);
+
+            if (ctx->tag) {
+                add_value(ctx, static_cast<int8_t>(value));
+                ctx->tag = std::nullopt;
+            } else if (ctx->states.back() == STATE::READING_INDEF_ARRAY) {
+                ctx->byte_array.push_back(static_cast<std::byte>(value));
+                return;
+            } else if (ctx->states.back() == STATE::READING_ARRAY) {
+                ctx->byte_array.push_back(static_cast<std::byte>(value));
+                if (ctx->byte_array.size() == ctx->byte_array_length) {
+                    add_value(ctx, ctx->byte_array);
+                    ctx->states.pop_back();
+                } else {
+                    return;
+                }
+            } else {
+                add_value(ctx, value);
+            }
+
+            check_map_end(ctx);
+        }
+    };
+
+}
 
 std::ostream& operator<<(std::ostream& out, const cosim::serialization::node& data)
 {
-    constexpr int indentStep = 2;
-    print_tree(out, data, 0, indentStep);
+    auto root = cbor_new_indefinite_map();
+    serialize_cbor(root, data);
+
+    unsigned char* buffer;
+    size_t buffer_size;
+
+    cbor_serialize_alloc(root, &buffer, &buffer_size);
+    out.write(reinterpret_cast<const char*>(buffer), static_cast<std::streamsize>(buffer_size));
+
+    free(buffer);
+    cbor_decref(&root);
     return out;
 }
+
+
+std::istream& operator>>(std::istream& in, cosim::serialization::node& root)
+{
+    in.seekg(0, std::ios::end);
+    std::streamsize length = in.tellg();
+    in.clear();
+    in.seekg(0, std::ios::beg);
+    std::vector<uint8_t> buf(length);
+    in.read(reinterpret_cast<char*>(buf.data()), length);
+
+    // Initializing the reader context object
+    cbor_reader_ctx ctx{root};
+
+    struct cbor_callbacks cbs = cbor_empty_callbacks;
+    struct cbor_decoder_result decode_result{};
```

Comment:
You don't need to prefix these with `struct` in C++, that's a C thing.

Reply:
Fixed